feat(rpc): Allow filtering when we can't use VirtualColumnContexts

wmak opened the pull request:
- VirtualColumnContexts aren't available on timeseries, so if we filter by a vcc field (project, device.class) we need custom logic
- From a discussion in slack going to drop device.module support, will do that in a follow up PR

wmak committed:
feat(rpc): Allow filtering when we can't use VirtualColumnContexts

- VirtualColumnContexts aren't available on timeseries, so if we filter
  by a vcc field (project, device.class) we need custom logic
- From a discussion in slack going to drop device.module support, will
  do that in a follow up PR

wmak committed:
fix: mistake from typing

Zylphrex commented on src/sentry/search/eap/spans.py:
```diff
@@ -310,6 +310,61 @@ def _resolve_having(
+        context: VirtualColumnContext | None,
+    ) -> tuple[ResolvedColumn, float | datetime | Sequence[float] | Sequence[str]]:
+        raw_value = term.value.raw_value
+        if context is None or isinstance(raw_value, datetime):
```
Why do you need `isinstance(raw_value, datetime)`? Is something special about `datetime` values and columns requiring virtual column contexts?

Zylphrex commented on src/sentry/search/eap/spans.py:
```diff
@@ -310,6 +310,61 @@ def _resolve_having(
+        elif isinstance(raw_value, list):
+            new_value = []
+            for raw_iterable in raw_value:
+                if term.key.name not in constants.PROJECT_FIELDS and raw_iterable == "Unknown":
```
What columns can support `Unknown` as a value? Should this assumption be specified somewhere for all virtual column contexts? Or is this a `device.class` specific thing?

wmak replied:
specific to device.class, good callout i'll change the first condition to specifically be for the current column

Zylphrex commented on src/sentry/search/eap/spans.py:
```diff
@@ -310,6 +310,61 @@ def _resolve_having(
+                        )
+                    )
+                else:
+                    new_value.append(reversed_context[raw_iterable])
```
Is a `KeyError` possible here?

wmak replied:
Nah previous elif block will catch it `elif raw_iterable not in reversed_context:`

Zylphrex commented on src/sentry/search/eap/spans.py:
```diff
@@ -310,6 +310,61 @@ def _resolve_having(
+                    new_value.append(reversed_context[raw_iterable])
+            final_raw_value = new_value
+        elif raw_value in reversed_context:
+            final_raw_value = reversed_context[raw_value]
```
Is a `KeyError` possible here?

wmak replied:
🤔 wait am I missing something, won't L348 catch this
```python
        elif raw_value in reversed_context:
```

Zylphrex replied:
nope im just bad at reading

Zylphrex commented on src/sentry/search/eap/spans.py:
```diff
@@ -310,6 +310,61 @@ def _resolve_having(
+                )
+            )
+
+        # Convert the resolved_column to the underlying column
+        if term.key.name in constants.PROJECT_FIELDS:
+            resolved_column, _ = self.resolve_attribute("project.id")
+            if isinstance(final_raw_value, list):
+                final_raw_value = [int(val) for val in final_raw_value]
+            else:
+                final_raw_value = int(final_raw_value)
+        elif term.key.name == "device.class":
+            resolved_column, _ = self.resolve_attribute("sentry.device.class")
```
This will conflict with #83714 where you're trying to make the search resolver applicable to different datasets.

wmak replied:
😬 ugh right... I'll move this logic to the VCC definitions instead then and merge 83714 first

wmak committed:
ref: Address PR comments

- TODO: move all of this into VCC definition so that we can be dataset
  agnostic within the SearchResolver

wmak committed:
ref: move spans specific logic out

wmak committed:
fix: tests

wmak committed:
fix: typing

Zylphrex commented on src/sentry/search/eap/resolver.py:
```diff
@@ -305,15 +318,83 @@ def _resolve_having(
+        if isinstance(raw_value, list):
+            new_value = []
+            for raw_iterable in raw_value:
+                if context.default_value and context.default_value == raw_iterable:
+                    # Avoiding this for now, while this could work with the Unknown:"" mapping
+                    # But that won't work once we use the VirtualColumnContext.default_value
+                    raise InvalidSearchQuery(
+                        f"Using {raw_iterable} in an IN filter is not currently supported"
+                    )
```
Why is the default value special when used in an `IN` filter?

wmak replied:
good catch, i'll disable in both sides for now and re-enable when i fix

wmak committed:
fix: double replacing float

wmak committed:
mark filters for unknown as unimplemented for now
